src/cpp/shared_core/FileLogDestination.cpp

```diff
@@ -208,18 +208,36 @@ void FileLogDestination::writeLog(LogLevel in_logLevel, const std::string& in_me
       return;
 
    // Lock the mutex before attempting to write.
-   boost::unique_lock<boost::mutex> lock(m_impl->Mutex);
+   try
+   {
+      boost::lock_guard<boost::mutex> lock(m_impl->Mutex);
 
-   // Open the log file if it's not open. If it fails to open, log nothing.
-   if (!m_impl->LogOutputStream && !m_impl->openLogFile())
-      return;
+      // Open the log file if it's not open. If it fails to open, log nothing.
+      if (!m_impl->openLogFile())
+         return;
 
-   // Rotate the log file if necessary. If it fails to rotate, log nothing.
-   if (!m_impl->rotateLogFile())
-      return;
+      // Rotate the log file if necessary. If it fails to rotate, log nothing.
+      if (!m_impl->rotateLogFile())
+         return;
 
-   (*m_impl->LogOutputStream) << in_message;
-   m_impl->LogOutputStream->flush();
+      (*m_impl->LogOutputStream) << in_message;
+      m_impl->LogOutputStream->flush();
+
+      // If the output stream has bad state after writing, it might have been closed. Try re-opening it and writing the
+      // message again. Often it is not possible to tell that a stream has failed until a write is attempted.
+      if (!m_impl->LogOutputStream->good())
+      {
+         if (!m_impl->openLogFile())
+            return;
+
+         (*m_impl->LogOutputStream) << in_message;
```

Comment:
Is there ever a case where we could end up duplicating a message in the log output here? (Is it worth worrying about? Seems like maybe not since this should be a rare error case in theory at least)

Reply:
I considered this - I think it's technically possible that some or all of a message is written before the stream fails and the `badbit` is set, but in my opinion it would be better to have the message logged twice than to be in the situation where the message (and subsequent messages) isn't logged at all.

I also tried testing the streams goodness by writing `""` to the stream, but that is a no-op it appears. I'd also rather have one log appear twice than have some non-log data appear in the stream every time we are going to log something.